Integrate IBM watsonx to Inference API for text embeddings

saikatsarkar056 opened the pull request:
This PR adds [IBM watsonx AI](https://dataplatform.cloud.ibm.com/docs/content/wsj/analyze-data/fm-embed-overview.html?context=wx) text embeddings support for the [inference APIs](https://www.elastic.co/guide/en/elasticsearch/reference/current/inference-apis.html).

### Prerequisites to Model Creation
- An IBM Account with reservation to watsonx AI. Please read this [doc](https://docs.google.com/document/d/1_zn2Qz_CPJJHnma5BNQ4lbVOh-bvi55LKGqubd-SYTg/edit#heading=h.ivczzpjehg2u) on how to get access.
- An API key used to access IBM cloud

### Inference Model Creation:

```
PUT _inference/text_embedding/{inference_model_id}
{
   "service": "watsonxai",
    "service_settings": {
        "api_key": "<<API_KEY>>",
        "url": "<<URL_TO_IBM_CLOUD>>",
        "model_id": "<<MODEL_ID>>",
        "project_id": "<<PROJECT_ID>>",
        "api_version": "<<API_VERSION>>"
    }
}
```

### Inference Model Inference
```
POST _inference/text_embedding/{inference_model_id}
{
  "input": <single string, or array of strings>
}
```

# Manual test
Following steps have been taken to test watsonx ai integration
- Create an inference endpoint using watsonx ai
- Generate text_embeddings using the inference endpoint
- Create an index
- Create a semantic_text field for the new index and refer to watson ai inference endpoint
- Verify that embeddings are generated for the semantic_text field while inserting data to the index

Please find the testing in the following screen-recording:

https://github.com/user-attachments/assets/38cc0573-644d-49e9-b184-97f5676c4f71

elasticsearchmachine commented:
Pinging @elastic/search-experiences-team (Team:Search - Experiences)

elasticsearchmachine commented:
Pinging @elastic/ent-search-eng (Team:SearchOrg)

saikatsarkar056 commented:
@timgrein, @prwhelan, @jonathan-buttner 

I’m currently working on adding some tests. Would it be okay to add the tests in a separate PR to keep the PR size smaller? or, Do you suggest the tests in the current PR?

kderusso commented:
My feedback is to add tests in this PR, to ensure that the code we're committing is solid.

maxhniebergall commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/ibmwatsonx/IbmWatsonxResponseHandler.java:
```diff
@@ -0,0 +1,68 @@
+        } else if (statusCode >= 300 && statusCode < 400) {
+            throw new RetryException(false, buildError(REDIRECTION, request, result));
+        } else {
+            throw new RetryException(false, buildError(UNSUCCESSFUL, request, result));
```
@prwhelan, @davidkyle, do you know why we aren't specifically catching 400 errors here? I see this matches our other integrations, but I'm not sure why we're doing it this way. I can't see us catching them anywhere else. I guess they get rolled into this UNSUCCESSFUL, but wouldn't a BAD_REQUEST error message be better?

maxhniebergall commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/request/ibmwatsonx/IbmWatsonxEmbeddingsRequestEntity.java:
```diff
@@ -0,0 +1,39 @@
+
+public record IbmWatsonxEmbeddingsRequestEntity(List<String> inputs, String modelId, String projectId) implements ToXContentObject {
+
+    private static final String INPUTS_FIELD = "inputs";
```
Do we not need the API version here?

saikatsarkar056 replied:
API version is not part of the request body rather it is sent as part of the url param. This is why API version is not added here.

maxhniebergall commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/ServiceFields.java:
```diff
@@ -18,6 +18,8 @@ public final class ServiceFields {
     public static final String MAX_INPUT_TOKENS = "max_input_tokens";
     public static final String URL = "url";
     public static final String MODEL_ID = "model_id";
+    public static final String PROJECT_ID = "project_id";
```
Nit: Does this need to be in the ServiceFields? It seems like this is the only service that uses ProjectID, so it's not really a shared field

saikatsarkar056 replied:
I moved the project_id from ServiceFields to IbmWatsonxFields.

maxhniebergall reviewed (commented):
LGTM, but I would like to see the tests before we merge

saikatsarkar056 committed:
Log the exception if  Bearer token generation fails

saikatsarkar056 committed:
Set rate limit

saikatsarkar056 committed:
Add tests

saikatsarkar056 committed:
Apply spotless

saikatsarkar056 committed:
Add test for ServiceSettings

saikatsarkar056 committed:
Add test for EmbeddingsRequestEntity

saikatsarkar056 committed:
Add test for IbmWatsonxEmbeddingsRequestEntity

saikatsarkar056 committed:
Apply spotless

saikatsarkar056 committed:
Add tests for IbmWatsonxEmbeddingsResponseEntity

saikatsarkar056 committed:
Fix the issue with long line

saikatsarkar056 committed:
Fix tests for IbmWatsonxEmbeddingsActionTests

saikatsarkar056 committed:
Apply spotless

saikatsarkar056 committed:
Move project_id from ServiceFields to IbmWatsonxServiceFields

saikatsarkar056 commented:
@maxhniebergall, @timgrein, @prwhelan I added the tests and addressed the feedbacks. It would be great if you please review the PR now. Thank you for help.

saikatsarkar056 committed:
Check 400 Bad Request

saikatsarkar056 committed:
Avoid logging exception since this may contain the bearer token

saikatsarkar056 commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/request/ibmwatsonx/IbmWatsonxRequest.java:
```diff
@@ -0,0 +1,64 @@
+                return (String) map.get("access_token");
+            });
+        } catch (Exception e) {
+            logger.error("Failed to add Bearer token to the request", e);
```
I changed this part to omit logging the exception since the exception may contain access token.

@prwhelan What is the best way to plumb the error back to the user? Should we throw another exception from the catch block?

saikatsarkar056 replied:
> I assume if we send an empty bearer token, this will ultimately fail the request with a 401 or 403 from IBM, whereas the actual error comes from this call and could either be a 500 if we made a mistake or another 401/403 if the API Key was wrong or something.

If the bearer token is empty, the user will receive 401 authentication error

prwhelan replied:
> @prwhelan What is the best way to plumb the error back to the user? Should we throw another exception from the catch block?

Yeah that would be fine, an exception thrown here will be caught in https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/http/retry/RetryingHttpSender.java#L127

We may want to reuse the response handler to make the distinction between a retryable exception and non-retryable, but that can also be a later change

saikatsarkar056 replied:
I updated the code to throw an exception now.

saikatsarkar056 commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/http/sender/IbmWatsonxRequestManager.java:
```diff
@@ -0,0 +1,27 @@
+        public static RateLimitGrouping of(IbmWatsonxModel model) {
+            Objects.requireNonNull(model);
+
+            return new RateLimitGrouping(model.rateLimitServiceSettings().modelId().hashCode());
```
I researched how to establish a 1-to-1 mapping between the model ID and plan ID but couldn’t find any documentation providing that information. According to the [plan details](https://www.ibm.com/docs/en/watsonx/saas?topic=learning-watson-machine-plans#wml-plan-details), different plans offer varying capacities; however, no specific plan restricts a user to use any particular model. The [plan documentation](https://www.ibm.com/docs/en/watsonx/saas?topic=learning-watson-machine-plans#wml-plan-details) states that the Lite plan has the lowest limit of 2 requests per second. Based on this, I set the [default rate limit](https://github.com/elastic/elasticsearch/pull/111770/files#diff-e3bae9942ccc035b943c9b8a6ab15092d984321673f72daac19196f9ef9b038fR57) for Watsonx to 120 requests per minute (2 requests per second).

prwhelan replied:
Yeah the documentation doesn't seem to indicate how requests are associated with billing or usage - we may have to change from model id to something like api key, but that can be a later change

maxhniebergall replied:
It seems like a plan ID is like an account ID? But we don't have access to these? I tend to agree that we should be rate limiting on API key in that case, rather than model ID.

saikatsarkar056 replied:
It would be great if we can address this in a separate PR. WDYT?

saikatsarkar056 commented:
@maxhniebergall, @prwhelan Thank you for the feedbacks. I tried to address all the feedbacks. It would be great if you please provide another round of reviews.

maxhniebergall reviewed (commented):
Hi Saikat, this LGTM, just a couple questions

Why is this marked as a non-issue instead of an enhancement?

Can you describe the manual testing process you did to verify the real world behaviour against IBM servers?

Great work!

elasticsearchmachine commented:
Hi @saikatsarkar056, I've created a changelog YAML for you.

saikatsarkar056 commented:
> Hi Saikat, this LGTM, just a couple questions
> 
> Why is this marked as a non-issue instead of an enhancement?
> 
> Can you describe the manual testing process you did to verify the real world behaviour against IBM servers?
> 
> Great work!

I added the steps and a screen-recording for the manual testing. Also, I added the **enhancement** label instead of non-issue.

saikatsarkar056 committed:
Throw an exception if the creation of Bearer token fails

saikatsarkar056 commented:
@maxhniebergall, @prwhelan The feedbacks have been addressed. Thank you for help!

elasticsearchmachine commented:
Hi @saikatsarkar056, I've created a changelog YAML for you.

saikatsarkar056 commented:
@mark-vieira This change is targeted for v8.16 release. Do you know why we need to use v9.0.0 label instead of v8.16 for this PR? Thank you for help.

mark-vieira commented:
> @mark-vieira This change is targeted for v8.16 release. Do you know why we need to use v9.0.0 label instead of v8.16 for this PR? Thank you for help.

The `main` branch has been bumped to version `9.0`. If this needs to be included in 8.16 then add the `v8.16.0` version label and backport to the `8.x` branch.

saikatsarkar056 commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/request/ibmwatsonx/IbmWatsonxRequest.java:
```diff
@@ -0,0 +1,65 @@
+                return (String) map.get("access_token");
+            });
+        } catch (Exception e) {
+            throw new XContentParseException("Failed to add Bearer token to the request");
```
@prwhelan Can you please suggest whether I have used the correct Exception here? Since we are using XContentType, I thought XContentParseException would be the appropriate exception. Let me know whether we need to use a different exception.

prwhelan replied:
That sounds reasonable - either that or `ElasticsearchStatusException` if we want to add a status code that is not 500 to it, like if they get a 4xx from an invalid / unauthorized api key

prwhelan replied:
I guess that code wouldn't go here. We'd add parsing above for that - reading the response error code like we do in `IbmWatsonxResponseHandler`.

saikatsarkar056 replied:
Updated

saikatsarkar056 commented on docs/changelog/111770.yaml:
```diff
@@ -0,0 +1,5 @@
+summary: Integrate IBM watsonx to Inference API for text embeddings
+area: Experiences
+type: enhancement
+issues: []
```
@maxhniebergall, @prwhelan What should we put in the issues field?

maxhniebergall replied:
you can leave it as it is.

mark-vieira replied:
The issues filed is used if this pull request closes an existing open issue. Most of the time these are bug fixes. In this case you can leave it empty.

saikatsarkar056 committed:
Throw exception based on the status code for generating Bearer token

saikatsarkar056 committed:
Revert "Throw exception based on the status code for generating Bearer token"

This reverts commit f3cd615b8eee1f39536175dee7fd4588a691f319.

saikatsarkar056 committed:
Delete .java-version file

saikatsarkar056 committed:
Fix test

saikatsarkar056 commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/request/ibmwatsonx/IbmWatsonxRequest.java:
```diff
@@ -0,0 +1,65 @@
+            httpPostForBearerToken.setHeader(HttpHeaders.CONTENT_TYPE, "application/x-www-form-urlencoded");
+
+            bearerToken = SocketAccess.doPrivileged(() -> {
+                HttpResponse response = httpClient.execute(httpPostForBearerToken);
```
@prwhelan

I tried to [add response validation](https://github.com/elastic/elasticsearch/pull/111770/commits/f3cd615b8eee1f39536175dee7fd4588a691f319) based on the status code, similar to error handling in the [IbmWatsonxResponseHandler](https://github.com/elastic/elasticsearch/pull/111770/files#diff-107f5e3d773721875e1240bb966dcbd38d2cee7b844b80355393236e881c2e86R50), where exceptions are thrown for specific status codes. However, this change requires appending `throws IOException` to createHttpRequest(), which affects several test files related to other providers.

To avoid affecting unrelated tests, can we address this validation in a separate PR since this one is specific to IBM Watsonx? I’ll create a GitHub ticket to track this and include the validation in a future PR. WDYT?

prwhelan replied:
I'm still worried that a user will get an unactionable and misleading error from an authentication issue that they can resolve themselves if they get the appropriate error message.

Why not combine parts of what you have in both commits?
https://github.com/prwhelan/elasticsearch/commit/b1e67eb0ead09f8a781c5666e9e54cd110ee4da7

saikatsarkar056 committed:
Update docs/changelog/111770.yaml

elasticsearchmachine commented:
Hi @saikatsarkar056, I've created a changelog YAML for you.

prwhelan commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/request/ibmwatsonx/IbmWatsonxEmbeddingsRequest.java:
```diff
@@ -0,0 +1,79 @@
+        httpPost.setEntity(byteEntity);
+        httpPost.setHeader(HttpHeaders.CONTENT_TYPE, XContentType.JSON.mediaType());
+
+        IbmWatsonxRequest.decorateWithBearerToken(httpPost, model.getSecretSettings());
```
```suggestion
        IbmWatsonxRequest.decorateWithBearerToken(httpPost, model.getSecretSettings(), model.getInferenceEntityId());
```

saikatsarkar056 replied:
Updated

saikatsarkar056 committed:
Use IOException instead of Exception

saikatsarkar056 committed:
Fix the tests

saikatsarkar056 committed:
Add end-to-end test and infer test

saikatsarkar056 committed:
Resolve merge conflicts

saikatsarkar056 committed:
Resolve merge conflicts

saikatsarkar056 committed:
Resolve merge conflicts

saikatsarkar056 commented:
@prwhelan I noticed that some tests, particularly for [inference](https://github.com/elastic/elasticsearch/blob/9df99f69c7d28e555564d53bd9bc92f57c4bff0e/x-pack/plugin/inference/src/test/java/org/elasticsearch/xpack/inference/services/googlevertexai/GoogleVertexAiServiceTests.java#L557) and [end-to-end testing](https://github.com/elastic/elasticsearch/blob/9df99f69c7d28e555564d53bd9bc92f57c4bff0e/x-pack/plugin/inference/src/test/java/org/elasticsearch/xpack/inference/external/action/googlevertexai/GoogleVertexAiEmbeddingsActionTests.java#L68) of GoogleVertexAI, were added to a notebook instead of the test files. This was done to avoid issues with the Bearer token.

Following the same approach, I omitted similar tests in this [commit](https://github.com/elastic/elasticsearch/pull/111770/commits/5fb73965a79c61428ab439a0eef2c0930ff4f19e). We can address the end-to-end and inference tests in one of the following ways:
•	Write the tests in a notebook
•	Mock the Bearer token generation API
